synapse/config/workers.py

```diff
@@ -26,6 +27,16 @@
 )
 from .server import ListenerConfig, parse_listener_def
 
+logger = logging.Logger(__name__)
+
+# an object to pass to the "default" parameter, with which we can
+# `is` against to be sure we have an undefined config option.
+UNDEFINED = object()
+
+# an object to pass to self.worker_to_update_user_directory if
+# update_user_directory was defined, is used to compare with 'is'.
+ANY_USER_DIRECTORY_WORKER = object()
```

Comment:
can we not avoid the need for this ugliness by deciding whether we should update the user directory in this module whether we should update the user dir (in the same way that we do with `run_background_tasks`, `start_pushers`, `send_federation`, etc)?

It feels like we've invented a completely different wheel when we already have a perfectly functional one.

Reply:
You mean something along the lines of checking for `worker_app` here if `update_user_directory` is defined, and then setting `update_user_directory_on = worker_name` depending on that?

Reply:
no, I mean getting rid of `config.worker.update_user_directory_on` altogether, and just having a boolean `config.worker.update_user_directory`, in much the same way as we have a `config.worker.send_federation`.

Reply:
That discussion is out of scope for this PR, please comment on https://github.com/matrix-org/synapse/issues/10339 with that, as this PR is applying the suggestions made there.

Reply:
I can't see anything about the implementation in #10339? It seems to just be talking about the format of the config file (which I'm not suggesting you change).

Reply:
https://github.com/matrix-org/synapse/issues/10339 talks about the disposition that we can't be sure the user directory will be updated at any time, I suggested to make the config value accept None to indicate explicitly that the user directory will be updated on the main process, and any string value for any worker name.

@anoadragon453 approved of this at the time, this PR is executing what is said in that issue, if I were to move ahead on this PR, then I need to know the consensus, so please discuss on that issue if you think this is not a good idea.

Reply:
@ShadowJonathan I think Rich is not referring to making any config format change, but instead that the config object only stores true/false whether **this** worker will run them.

I refer to `run_background_tasks` (one of the fields Rich mentioned):
```python
        background_tasks_instance = config.get("run_background_tasks_on") or "master"
        self.run_background_tasks = (
            self.worker_name is None and background_tasks_instance == "master"
        ) or self.worker_name == background_tasks_instance
```

Put in other words:
A Synapse process doesn't care exactly *who (which process)* is updating the directory. It only cares whether *it's* the one doing it or not.

Reply:
That snippet has the same behaviour as this current config change, it doesn't test for true/false, are you sure you got the right snippet?

Reply:
With some private clarification from rich (after i saw we were miscommunicating on this issue), i'll alter the internal variable to `update_user_directory: bool`, which is derived from the yaml config value `update_user_directory_on: Option[str]`, with backwards compat from the yaml config option `update_user_directory`